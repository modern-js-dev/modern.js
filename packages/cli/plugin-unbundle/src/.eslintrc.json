{
  "rules": {
    "no-param-reassign": "off",
    "@typescript-eslint/no-loop-func": "off",
    "prefer-destructuring": "off",
    "consistent-return": "off",
    "max-statements": "off",
    "@typescript-eslint/no-shadow": "off",
    "max-depth": "off",
    "@typescript-eslint/member-ordering": "off",
    "max-lines": "off",
    "@typescript-eslint/no-base-to-string": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "callback-return": "off",
    "no-cond-assign": "off",
    "react-hooks/rules-of-hooks": "off"
  }
}
