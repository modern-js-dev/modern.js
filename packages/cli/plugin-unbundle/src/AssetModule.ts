import { path } from '@modern-js/utils';
import { cleanUrl } from './utils';

export const idToModules = new Map<string, AssetModule>();
export const fileToModules = new Map<string, AssetModule>();
export const urlToModules = new Map<string, AssetModule>();

export class AssetModule {
  // cleaned url
  url: string;

  // url with query or hash
  rawUrl: string;

  // resolved id from url
  _id: string;

  // absolute fs file path
  _filePath: string;

  // dependency's id
  dependencies: Set<string>;

  // depedent's id
  dependents: Set<string>;

  // import.hot.accept(deps, () => {}), deps id
  acceptIds: Set<string>;

  // import.hot.accept(() => {})
  selfAccepted: boolean;

  transformed: string | null;

  hmrTimestamp: number;

  etag: string;

  type: string;

  constructor(url: string) {
    this.rawUrl = url;
    this.url = cleanUrl(url);

    urlToModules.set(this.url, this);

    this._id = '';
    this._filePath = '';
    this.acceptIds = new Set<string>();
    this.dependencies = new Set<string>();
    this.dependents = new Set<string>();
    this.selfAccepted = false;
    this.transformed = null;
    this.hmrTimestamp = 0;
    this.etag = '';
    this.type = '';
  }

  get filePath() {
    return this._filePath;
  }

  set filePath(file: string) {
    this._filePath = file;
    this.type = path.extname(this._filePath);
    if (!fileToModules.has(file)) {
      fileToModules.set(file, this);
    }
  }

  get id() {
    return this._id;
  }

  set id(id) {
    this._id = id;
    if (!idToModules.has(id)) {
      idToModules.set(id, this);
    }
  }

  update() {
    // TODO
  }
}

export const createAssetModule = (url: string): AssetModule => {
  const existing = urlToModules.get(cleanUrl(url));

  if (existing) {
    return existing;
  }

  return new AssetModule(url);
};

// invalidate transform cache
// mark as been hmr update by timestamp
export const invalidateAssetModule = (
  assetModule: AssetModule,
  timestamp: number,
) => {
  assetModule.hmrTimestamp = timestamp;
  assetModule.transformed = null;
};
