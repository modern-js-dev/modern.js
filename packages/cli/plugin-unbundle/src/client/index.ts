// css HMR
const styles = new Map();

export const removeStyle = id => {
  if (styles.has(id)) {
    document.head.removeChild(styles.get(id));
    styles.delete(id);
  }
};

export const updateStyle = (id, code) => {
  const style = styles.get(id);
  if (!style) {
    const styleEl = document.createElement('style');
    const codeEl = document.createTextNode(code);
    styleEl.type = 'text/css';
    styleEl.appendChild(codeEl);
    document.head.appendChild(styleEl);
    styles.set(id, styleEl);
  } else {
    style.innerHTML = code;
  }
};

const ERROR_OVERLAY_NAME = 'modern-js-error-overlay';

// check if has  error overlay
function hasErrorOverlay() {
  return document.querySelectorAll(ERROR_OVERLAY_NAME).length;
}
// clear react error overlay
function clearErrorOverlay() {
  document.querySelectorAll(ERROR_OVERLAY_NAME).forEach(el => el.remove());
}

// create react error overlay
function createErrorOverlay(data) {
  const ErrorOverlay = customElements.get(ERROR_OVERLAY_NAME);
  if (ErrorOverlay) {
    const overlay = new ErrorOverlay(data);
    clearErrorOverlay();
    document.body.appendChild(overlay);
  }
}

type LoggerFunction = (...args: any[]) => void;
type LoggerType = {
  info: LoggerFunction;
  error: LoggerFunction;
  warn: LoggerFunction;
};
const logger = [`info`, 'error', 'warn'].reduce((m, c) => {
  m[c] = (...params) =>
    // eslint-disable-next-line no-console
    console[c === 'info' ? 'log' : c](`[HMR]:`, ...params);
  return m;
}, {} as LoggerType);

let isFisrtUpdate = true;

const handleMessage = event => {
  const data = JSON.parse(event.data);

  switch (data.type) {
    case 'connected': {
      logger.info(`Connected`);
      // TODO: socket ping
      break;
    }
    case 'reload': {
      window.location.reload();
      break;
    }
    case 'update': {
      // if there is already a error in first update
      // which meanings some resource load failed
      // so should reload page to ensure app run correctly
      if (isFisrtUpdate && hasErrorOverlay()) {
        window.location.reload();
        return;
      }

      isFisrtUpdate = false;

      try {
        const timestamp = data.hmrTimestamp;
        for (const { id, accepted } of data.changes) {
          doAccept(id, accepted, timestamp);
        }
        clearErrorOverlay();
      } catch (err) {
        logger.error('Hot Update Error', err);
        createErrorOverlay({
          title: 'Hot Update Error',
          loc: {
            file: data.url || '',
          },
          message: err.message,
          stack: err.stack,
        });
      }

      break;
    }
    // handle removed dependencies after hmr update
    // eg: remove style tags in head element
    case 'prune': {
      for (const id of data.changes) {
        doPrune(id);
      }
      break;
    }
    case 'error': {
      logger.error(data.message);
      createErrorOverlay(data);
      break;
    }
    case 'close': {
      logger.info(`Disconnected!`);
      break;
    }
    default: {
      logger.error(`Invalid message type:`, data);
    }
  }
};

const handleError = err => {
  logger.error(`connection error`, err);
};

let retries = 0;
const handleClose = () => {
  logger.info(`Disconnected!`);
  if (retries <= 10) {
    retries += 1;
    setTimeout(() => {
      initSocket();
    }, 1000 * Math.pow(2, retries) + Math.random() * 100);
  }
};

// init websocket client connection
const initSocket = () => {
  const ws = new WebSocket(
    `${location.origin.replace(/http/, 'ws')}/_modern_js_hmr_ws`,
  );
  ws.onerror = handleError;
  ws.onmessage = handleMessage;
  ws.onclose = handleClose;
};

initSocket();

const strip = url => url.replace(/(\?.*$|#.*$)/, '');

const noop = function () {
  /** */
};

const modules = new Map();

let updateQueue = [];
let updating = false;

const doAccept = async (id, acceptedPath, timestamp) => {
  if (!updateQueue.includes(id)) {
    updateQueue.push({ id, acceptedPath });
  }

  if (!updating) {
    updating = true;
    await Promise.resolve();
    updating = false;
    const loading = updateQueue.map(({ id: path, acceptedPath: accepted }) =>
      update(path, accepted, timestamp),
    );
    updateQueue = [];
    await Promise.all(loading);
  }
};

const update = async (id, acceptedPath, timestamp) => {
  if (!modules.has(id)) {
    return;
  }

  const mod = modules.get(id);
  let callbacks = [];
  const modulesToUpdate = new Set();
  if (mod._selfAccepted) {
    modulesToUpdate.add(id);
    callbacks = [mod._selfAccepted];
  } else {
    modulesToUpdate.add(id);
    mod._acceptedCallbacks.forEach(({ deps, callback }) => {
      if (deps.some(dep => dep === acceptedPath)) {
        modulesToUpdate.add(acceptedPath);
        callbacks.push(callback);
      }
    });
  }

  const loadedModules = await Promise.all(
    Array.from(modulesToUpdate).map(m => {
      const newUrl = __addQuery__(m, `t=${timestamp}`);
      return import(newUrl);
    }),
  );

  for (const callback of callbacks) {
    typeof callback === 'function' && callback(loadedModules);
  }

  logger.info(`hot updated: accepted via ${id}`);
};

/** Called when a new module is loaded, to run cleanup on the old module (if needed) */
const doDispose = id => {
  const mod = modules.get(id);
  const callbacks = mod._disposeCallbacks;

  callbacks.map(callback => callback());
};

const doPrune = id => {
  if (modules.has(id)) {
    const mod = modules.get(id);
    const callbacks = mod._prunCallbacks;
    callbacks.map(callback => callback());
    modules.delete(id);
  }
};

export const createHotContext = url => {
  const id = strip(url);
  const mod = modules.get(id);

  if (mod) {
    // mod.lock();
    doDispose(mod._id);
    mod._disposeCallbacks = [];
    mod._acceptedCallbacks = [];
    mod._selfAccepted = false;
    return mod;
  }

  const hot = {
    // private
    _id: id,
    _declinedDeps: {},
    _acceptedDeps: {},
    _selfAccepted: false,
    _selfDeclined: false,
    _disposeCallbacks: [],
    _acceptedCallbacks: [],
    _prunCallbacks: [],
    // hot API
    accept(dep, callback) {
      if (dep === undefined) {
        hot._selfAccepted = true;
      } else if (typeof dep === 'function') {
        hot._selfAccepted = dep;
      } else if (Array.isArray(dep)) {
        hot._acceptedCallbacks.push({ deps: dep, callback: callback || noop });
      } else {
        hot._acceptedCallbacks.push({
          deps: [dep],
          callback: callback || noop,
        });
      }
    },
    decline(dep) {
      if (dep === undefined) {
        hot._selfDeclined = true;
      } else if (Array.isArray(dep)) {
        for (const d of dep) {
          hot._acceptedDeps[d] = true;
        }
      } else {
        hot._declinedDeps[dep] = true;
      }
    },
    dispose(callback) {
      hot._disposeCallbacks.push(callback);
    },
    prune(callback) {
      hot._prunCallbacks.push(callback);
    },
    invalidate() {
      window.location.reload();
    },
  };

  modules.set(id, hot);

  return hot;
};

export const __addQuery__ = (url, query) =>
  /\?.*/.test(url) ? `${url}&${query}` : `${url}?${query}`;
