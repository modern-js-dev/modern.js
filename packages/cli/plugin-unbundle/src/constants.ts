import path from 'path';

export const MODE = 'development';

export const WEB_MODULES_DIR = './node_modules/.modern_js_web_modules';

export const META_DATA_FILE_NAME = 'metadata.json';

export const DEFAULT_DEPS: string[] = [
  'react',
  'react-dom',
  '@modern-js/runtime',
  '@modern-js/runtime/loadable',
  '@modern-js/runtime/model',
  '@modern-js/runtime/styled',
  '@modern-js/runtime/head',
  '@modern-js/runtime/model',
  '@modern-js/runtime/plugins',
  '@modern-js/runtime/router',
  '@modern-js/runtime/ssr',
  '@modern-js/create-request',
];

export const MODERN_JS_INTERNAL_PACKAGES: Record<string, string> = {
  '@modern-js/plugin-state': '@modern-js/runtime',
  '@modern-js/plugin-router': '@modern-js/runtime',
  '@modern-js/plugin-ssr': '@modern-js/runtime',
  '@modern-js/create-request/client': '@modern-js/plugin-unbundle',
};

export const VIRTUAL_DEPS_MAP: Record<string, string> = {
  '@modern-js/create-request': `export * from '@modern-js/create-request/client';`,
  '@modern-js/runtime/plugins': `
    export { default as router } from '@modern-js/plugin-router';
    export { default as ssr } from '@modern-js/plugin-ssr';
    export { default as state } from '@modern-js/plugin-state';
  `,
  '@modern-js/runtime/router': `export * from '@modern-js/plugin-router';`,
  '@modern-js/runtime/model': `export * from '@modern-js/plugin-state';`,
  '@modern-js/runtime/ssr': `export * from '@modern-js/plugin-ssr';`,
  domino: 'export {}; export default {};',
};

export const HOST = '0.0.0.0';

// [@\w] - Match a word-character or @ (valid package name)
// (?!.*(:\/\/)) - Ignore if previous match was a protocol (ex: http://)
export const BARE_SPECIFIER_REGEX = /^[@\w](?!.*(:\/\/))/;

export const DEFAULT_EXTENSIONS = [
  'web.mjs',
  'mjs',
  'web.js',
  'js',
  'web.ts',
  'ts',
  'web.tsx',
  'tsx',
  'json',
  'web.jsx',
  'jsx',
].map(ext => `.${ext}`);

export const CSS_REGEX = /\.(css|sass|scss|less)($|\?)/;
export const CSS_MODULE_REGEX = /\.module.(css|sass|scss|less)($|\?)/;

// hmr sockjs client, should inject modern build into client
export const DEV_CLIENT_PATH = path.resolve(__dirname, '../modern/client');

export const DEV_CLIENT_PATH_ALIAS = `@_modern_js_dev_client`;

export const DEV_CLIENT_URL = `/${DEV_CLIENT_PATH_ALIAS}/index.js`;

export const ASSETS_REGEX =
  /\.(gif|png|jpe?g|webp|bmp|ico|svg|woff|woff2|eot|ttf|otf|ttc)$/;

export const IS_DISABLE_REACT_REFRESH = process.env.FAST_REFRESH === 'false';

export const CSS_URL_FUNCTION_REGEX =
  /url\(\s*('[^']+?'|"[^"]+?"|[^'"]+?)\s*\)/g;

export const GLOBAL_CACHE_DIR_NAME = '__modern_js_web_modules__';

// cjs node_modules to esm node_modules
export const TEMP_MODULES_DIR = './node_modules/.cache/modern_js_modules';

// babel macros extensions
export const BABEL_MACRO_EXTENSIONS = '.macro';

// api function file query
export const LAMBDA_API_FUNCTION_QUERY = '__lambda_api_function__';

// dependencies that should be ignored in deps optimizaion
// can't convert to esm format
export const IGNORE_OPTIMIZE_DEPS = ['domino'];

// default lazy import ui components library
export const DEFAULT_LAZY_IMPORT_UI_COMPONENTS = [
  'antd',
  '@arco-design/web-react',
];

export const ESBUILD_RESOLVE_PLUGIN_NAME = 'esm-resolve-plugin';

export const BFF_API_DIR = './api';

export const DEFAULT_PDN_HOST = `pdn.zijieapi.com`;
