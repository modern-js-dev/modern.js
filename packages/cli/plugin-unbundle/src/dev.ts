import { chalk, logger } from '@modern-js/utils';
import { IAppContext, NormalizedConfig } from '@modern-js/core';

export const startTimer = {
  start: 0,
  end: 0,
};

export const dev = async (
  config: NormalizedConfig,
  appContext: IAppContext,
) => {
  logger.log(chalk.cyanBright(`Starting the development server...\n`));

  startTimer.start = Date.now();

  const [{ startDevServer }, { createHtml }] = await Promise.all([
    import('./server'),
    import('./create-entry'),
  ]);

  // create html template for each js entry
  createHtml(config, appContext);

  const closeServer = await startDevServer(config, appContext);
  return closeServer;
};
