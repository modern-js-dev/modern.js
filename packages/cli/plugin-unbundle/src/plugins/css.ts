import fs from 'fs';
import path from 'path';
import { chalk, signale as logger } from '@modern-js/utils';
import { Alias } from '@rollup/plugin-alias';
import { Plugin as RollupPlugin, SourceMap } from 'rollup';
import postcss, { AcceptedPlugin, ProcessOptions } from 'postcss';
import { codeFrameColumns } from '@babel/code-frame';
import type { LegacyImporterResult } from 'sass';
import less from 'less';
import type { IAppContext, NormalizedConfig } from '@modern-js/core';
import {
  getLessConfig,
  getPostcssConfig,
  getSassConfig,
} from '@modern-js/css-config';
import { isCSSRequest, replaceAsync, addQuery, hasDependency } from '../utils';
import {
  CSS_MODULE_REGEX,
  CSS_URL_FUNCTION_REGEX,
  DEV_CLIENT_URL,
} from '../constants';
import { HMRError } from '../websocket-server';
import { createAssetModule, fileToModules } from '../AssetModule';
import { normalizeAlias } from './alias';

export interface PreProcessOptions {
  lessOptions?: Record<string, any>;
  sassOptions?: Record<string, any>;
  additionalData?: string | ((content: string, filename: string) => string);
}

export interface TransformRes {
  css?: string;
  errors?: HMRError[];
  deps?: string[];
  map?: SourceMap;
}

let processorOptions: {
  less: PreProcessOptions;
  sass: PreProcessOptions;
};

let nodeModulesPaths: Array<string> = [];

let postcssConfig: {
  postcssOptions: ProcessOptions & {
    plugins: AcceptedPlugin[];
  };
};

let tailwindConfig: any;

// record css modules locals json
const moduleLocalsMap = new Map();

const initProcessorOptions = (config: NormalizedConfig) => {
  if (processorOptions) {
    return;
  }

  const lessOptions = getLessConfig(config);
  const sassOptions = getSassConfig(config);

  processorOptions = {
    less: lessOptions,
    sass: sassOptions,
  };
};

const initTailwindConfig = (config: NormalizedConfig) => {
  try {
    if (!tailwindConfig) {
      // TODO: talwindcss config.
      tailwindConfig = (config.tools as any).tailwind;
    }
  } catch (err: any) {
    if (err.code !== 'MODULE_NOT_FOUND') {
      throw err;
    }
    // ignore err
  }
};

const initPostcssConfig = (
  config: NormalizedConfig,
  { appDirectory }: IAppContext,
) => {
  if (postcssConfig) {
    return;
  }

  // FIXME: type any
  postcssConfig = getPostcssConfig(appDirectory, config, false) as any;
};

let matchAlias: (request: string) => Alias | undefined;

const createIsAliasRequest = (
  config: NormalizedConfig,
  appContext: IAppContext,
  defaultDeps: string[],
) => {
  if (!matchAlias) {
    const aliasOptions = normalizeAlias(config, appContext, defaultDeps);
    matchAlias = (request: string) =>
      aliasOptions.find((alias: Alias) => {
        const { find } = alias;
        if (find instanceof RegExp) {
          return find.test(request);
        } else if (typeof find === 'string') {
          if (request.length < find.length) {
            return false;
          }
          if (request.length === find.length) {
            return true;
          }

          const requestStartsWithKey = request.startsWith(find);
          const requestHasSlashAfterKey = request
            .substring(find.length)
            .startsWith('/');
          return requestStartsWithKey && requestHasSlashAfterKey;
        }
        return false;
      });
  }
};

function getSource(
  source: string,
  filename: string,
  additionalData?: string | ((content: string, filename: string) => string),
) {
  if (!additionalData) {
    return source;
  }
  if (typeof additionalData === 'function') {
    return additionalData(source, filename);
  }
  return `${additionalData}\n${source}`;
}

const isCSSModule = (config: NormalizedConfig, id: string): boolean => {
  if (id.includes(`node_modules`)) {
    return false;
  }

  const {
    output: { disableCssModuleExtension },
  } = config;

  if (disableCssModuleExtension) {
    return true;
  }

  return CSS_MODULE_REGEX.test(id);
};

export const hasTailwind = (appDirectory: string) =>
  hasDependency(appDirectory, `@modern-js/plugin-tailwindcss`) &&
  hasDependency(appDirectory, `tailwindcss`);

export class CustomLessFileManager extends less.FileManager {
  async loadFile(
    filename: string,
    currentDirectory: string,
    options: Less.LoadFileOptions,
    environment: Less.Environment,
  ) {
    let result;

    if (filename.startsWith('~')) {
      filename = filename.slice(1);
    }

    try {
      result = await super.loadFile(
        filename,
        currentDirectory,
        options,
        environment,
      );
    } catch (err: any) {
      if (err.type !== 'File') {
        return Promise.reject(err);
      }

      const matched = matchAlias(filename);
      if (matched) {
        return super.loadFile(
          filename.replace(matched.find, matched.replacement),
          currentDirectory,
          options,
          environment,
        );
      } else {
        return Promise.reject(err);
      }
    }

    return result;
  }
}

const compileLess = async (
  code: string,
  filename: string,
  options: PreProcessOptions,
): Promise<TransformRes> => {
  const less = require('less');

  try {
    const res = await less.render(
      getSource(code, filename, options.additionalData),
      {
        filename,
        sourceMap: true,
        ...options.lessOptions,
        plugins: [
          {
            install(lessInstance: any, pluginManager: any) {
              pluginManager.addFileManager(new CustomLessFileManager());
            },
          },
        ],
      },
    );

    return { css: res.css, deps: res.imports || [], map: res.map };
  } catch (err: any) {
    if (err.filename) {
      err.frame = codeFrameColumns(fs.readFileSync(err.filename, 'utf8'), {
        start: {
          line: err.line,
          column: err.column,
        },
      });
    }

    err.loc = {
      file: err.filename,
      line: err.line,
      column: err.column,
    };

    return { errors: [err] };
  }
};

const compileSass = async (
  code: string,
  filename: string,
  options: PreProcessOptions,
): Promise<TransformRes> => {
  const sass = require('sass');

  try {
    const result: any = await new Promise((resolve, reject) => {
      sass.render(
        {
          ...options.sassOptions,
          data: getSource(code, filename, options.additionalData),
          identSyntax: filename.endsWith('.sass'),
          includePaths: [path.dirname(filename), ...nodeModulesPaths],
          outFile: filename,
          file: filename,
          sourceMap: true,
          importer(
            url: string,
            prev: string,
            done: (data: LegacyImporterResult) => void,
          ) {
            if (url.startsWith('~')) {
              url = url.slice(1);
            }
            const matched = matchAlias(url);
            done({
              file: matched
                ? url.replace(matched.find, matched.replacement)
                : url,
            });
          },
        },
        (err: Error | null, result: any) => {
          if (err) {
            reject(err);
          } else {
            resolve(result);
          }
        },
      );
    });

    return {
      css: result.css.toString(),
      deps: result.stats.includedFiles || [],
      map: result.map.toString(),
    };
  } catch (e: any) {
    e.loc = {
      file: e.file,
      line: e.line,
      column: e.column,
    };
    return { errors: [e] };
  }
};

const transformCSS = async (
  code: string,
  filename: string,
  config: NormalizedConfig,
  { appDirectory }: IAppContext,
): Promise<TransformRes> => {
  const extname = path.extname(filename);

  let css: string | undefined;
  let errors: HMRError[] | undefined;
  let deps: string[] | undefined = [];
  let map: SourceMap | undefined;
  // pre process
  switch (extname) {
    case '.less': {
      ({ css, errors, deps, map } = await compileLess(
        code,
        filename,
        processorOptions.less,
      ));
      break;
    }
    case '.sass':
    case '.scss': {
      ({ css, errors, deps, map } = await compileSass(
        code,
        filename,
        processorOptions.sass,
      ));
      break;
    }
    default: {
      // empty
    }
  }

  if (!deps) {
    deps = [];
  }

  if (errors?.length) {
    throw errors[0] as Error;
  }

  // less/sass 处理完之后空文件时，直接返回
  if (extname !== '.css' && !css) {
    return { css: '' };
  }

  // postcss
  const postcssPlugins = [...postcssConfig.postcssOptions.plugins];
  if (isCSSModule(config, filename)) {
    postcssPlugins.push(
      require('postcss-modules')({
        localsConvention: 'camelCase',
        generateScopedName: config.output.cssModuleLocalIdentName,
        globalModulePaths: [/\.global\.(css|scss|sass|less)$/],
        getJSON(_: string, _modules: Record<string, string>) {
          moduleLocalsMap.set(filename, _modules);
        },
      }),
    );
  }

  // As @modern-js/plugin-tailwindcss requires 'tailwindcss' as peer depenedency,
  // so we could safely require both plugins when using tailwind features
  if (hasTailwind(appDirectory)) {
    postcssPlugins.push(require('tailwindcss')(tailwindConfig));
  }

  try {
    const result = await postcss(postcssPlugins).process(css || code, {
      from: filename,
      to: filename,
      map: {
        inline: false,
        annotation: false,
        prev: map,
      },
      ...postcssConfig.postcssOptions,
    });

    if (result?.messages.length) {
      for (const message of result.messages) {
        if (message.type === 'dependency') {
          deps.push(message.file as string);
        } else if (message.type === 'warning') {
          let tips = `[${message.plugin as string}] ${message.text}\n`;
          if (message.line && message.column) {
            tips += `${codeFrameColumns(code, {
              start: {
                line: message.line,
                column: message.column,
              },
            })}`;
          }
          logger.warn(chalk.yellowBright(tips));
        }
      }
    }

    // deps tree for css  hmr
    const assetModule = fileToModules.get(filename);

    for (const dep of Array.from(new Set(deps))) {
      const depModule = createAssetModule(path.relative(filename, dep));
      depModule.filePath = dep;
      depModule.id = dep;
      assetModule!.dependencies.add(dep);
      depModule.dependents.add(assetModule!.id);
    }

    return {
      css: result.css,
      map: result.map as any,
    };
  } catch (err: any) {
    if (err.line !== undefined && err.column !== undefined) {
      err.loc = {
        file: err.file,
        line: err.line,
        column: err.column,
      };
      err.frame = codeFrameColumns(err.input.source, {
        start: {
          line: err.line,
          column: err.column,
        },
      });
    }
    throw err;
  }
};

// css url() assets rewrite
const rewriteCssUrl = async (
  code: string | undefined,
  replacer: (resource: string) => Promise<string>,
): Promise<string | undefined> => {
  if (!code) {
    return;
  }
  return replaceAsync(
    code,
    CSS_URL_FUNCTION_REGEX,
    async (match: string, $1: string) => {
      if (/^["']?(https?|data:|#)/.test($1.trim())) {
        return match;
      }
      const replaced = await replacer($1);
      return `url('${replaced}')`;
    },
  );
};

export const cssPlugin = (
  config: NormalizedConfig,
  appContext: IAppContext,
  defaultDeps: string[],
): RollupPlugin => ({
  name: 'esm-css',
  async transform(code: string, importer: string) {
    if (!isCSSRequest(importer)) {
      return null;
    }

    const { appDirectory } = appContext;

    // find parent node_modules paths for sass and less
    nodeModulesPaths = require('find-node-modules')({ cwd: appDirectory });

    initProcessorOptions(config);

    initPostcssConfig(config, appContext);

    initTailwindConfig(config);

    createIsAliasRequest(config, appContext, defaultDeps);

    const { css, map } = await transformCSS(code, importer, config, appContext);

    // css url() rewrite
    const rewrited = await rewriteCssUrl(
      css,
      async (resource: string): Promise<string> => {
        resource = resource.replace(/(^['"])|(['"]$)/g, '').replace(/^~/, '');

        // try to resolve url request
        // if resolved then return relative url
        // if can't resolve just return original resource
        // should ignore error here, such as cursor: url(mycursor.cur)
        try {
          const resolved = await this.resolve(resource, path.dirname(importer));

          const filePath =
            typeof resolved === 'object' ? resolved?.id || resource : resolved;

          const relativeUrl = path.relative(appDirectory, filePath);

          let rewrite = relativeUrl.startsWith('.')
            ? filePath
            : `/${relativeUrl}`;

          // resource HMR
          if (path.extname(filePath) === '.css') {
            rewrite = addQuery(rewrite, 'direct');
            const assetModule = fileToModules.get(importer);

            assetModule!.dependencies.add(rewrite);

            const depModule = createAssetModule(rewrite);
            depModule.filePath = filePath;
            depModule.id = rewrite;

            depModule.dependents.add(assetModule!.id);
          }

          return rewrite;
        } catch (err) {
          logger.warn(`Can't resolve resource ${resource} in ${importer}`);
          return resource;
        }
      },
    );

    const wrappedCss = [
      `import { updateStyle, removeStyle } from "${DEV_CLIENT_URL}"`,
      `const code = ${JSON.stringify(rewrited)}`,
      `const filename = ${JSON.stringify(importer)}`,
      `updateStyle(filename, code);`,
      moduleLocalsMap.has(importer) &&
        `export default ${JSON.stringify(moduleLocalsMap.get(importer))};`,
    ]
      .filter(Boolean)
      .join('\n');

    return { code: wrappedCss, map };
  },
});
