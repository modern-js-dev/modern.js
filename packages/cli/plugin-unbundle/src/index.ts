import type { CliPlugin } from '@modern-js/core';
import { PLUGIN_SCHEMAS } from '@modern-js/utils';
import { dev } from './dev';
import { hooks } from './hooks';

export * from './hooks';

export default (): CliPlugin => ({
  name: '@modern-js/plugin-unbundle',

  registerHook: hooks,

  setup: api => {
    let closeDevServer: (() => Promise<void>) | undefined;

    return {
      validateSchema() {
        return PLUGIN_SCHEMAS['@modern-js/plugin-unbundle'];
      },
      commands({ program }) {
        const appContext = api.useAppContext();
        const config = api.useResolvedConfigContext();

        const devCommand = program.commandsMap.get('dev');

        devCommand?.option('--unbundled', 'dev with unbundled mode');

        if (
          process.argv.slice(2)[0] === 'dev' &&
          process.argv.includes('--unbundled')
        ) {
          devCommand?.action(async () => {
            if (closeDevServer) {
              await closeDevServer();
              closeDevServer = undefined;
            }
            closeDevServer = await dev(api, config, appContext);
          });
        }
      },
      async beforeRestart() {
        if (closeDevServer) {
          await closeDevServer();
        }
        closeDevServer = undefined;
      },
      async htmlPartials({ entrypoint, partials }) {
        if (process.argv[2] === 'dev' && process.argv.includes('--unbundled')) {
          const appContext = api.useAppContext();
          const config = api.useResolvedConfigContext();

          const { createHtmlPartials } = await import('./create-entry');

          partials.head.push(
            createHtmlPartials(entrypoint, appContext, config),
          );
        }

        return {
          entrypoint,
          partials,
        };
      },
    };
  },
});
