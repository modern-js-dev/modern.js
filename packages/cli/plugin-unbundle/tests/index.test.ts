import * as path from 'path';
import { manager } from '@modern-js/core';
import plugin from '../src';

import { hasTailwind } from '../src/plugins/css';
import { dev } from '../src/dev';

jest.mock('../src/dev');
jest.mock('@modern-js/core', () => ({
  __esModule: true,
  ...jest.requireActual('@modern-js/core'),
  registerHook: jest.fn(),
}));

describe('plugin-unbundle', () => {
  let originalArgv: string[];
  beforeAll(() => {
    originalArgv = process.argv;
    process.argv = ['xxx/xxx/xxx/node', 'xxx/xxx/modern', 'dev', '--unbundled'];
  });

  afterAll(() => {
    process.argv = originalArgv;
  });

  it('default', () => {
    expect(plugin).toBeDefined();
  });

  it('test tailwind plugin exists', () => {
    expect(
      hasTailwind(path.join(__dirname, './fixtures/tailwind-example')),
    ).toBe(true);
  });

  it('schema', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const result = await runner.validateSchema();
    expect(result).toMatchSnapshot();
  });

  // eslint-disable-next-line max-statements
  it('commands', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const closeDevServerMock = jest.fn().mockReturnValue(Promise.resolve());
    jest.mocked(dev).mockResolvedValue(closeDevServerMock);
    const devCommand = {
      option: jest.fn(),
      action: jest.fn(),
    };
    const program = {
      commandsMap: {
        get: jest.fn().mockReturnValue(devCommand),
      },
    };
    await runner.commands({ program } as any);
    expect(program.commandsMap.get).toHaveBeenCalledWith('dev');
    expect(devCommand.option).toHaveBeenCalledWith(
      '--unbundled',
      expect.anything(),
    );
    expect(devCommand.action).toHaveBeenCalledTimes(1);
    const devCommandAction = devCommand.action.mock.calls[0][0];
    expect(devCommandAction).toBeTruthy();
    expect(typeof devCommandAction).toBe('function');
    await devCommandAction();
    expect(dev).toHaveBeenCalledTimes(1);
    expect(closeDevServerMock).not.toHaveBeenCalled();

    await runner.beforeRestart();
    await devCommandAction();
    expect(closeDevServerMock).toHaveBeenCalledTimes(1);

    closeDevServerMock.mockClear();
    await devCommandAction();
    expect(closeDevServerMock).toHaveBeenCalledTimes(1);
  });
});
