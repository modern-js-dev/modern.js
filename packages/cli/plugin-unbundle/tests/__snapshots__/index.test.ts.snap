// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-unbundle schema 1`] = `
Array [
  Array [
    Object {
      "schema": Object {
        "type": "boolean",
      },
      "target": "output.disableAutoImportStyle",
    },
    Object {
      "schema": Object {
        "properties": Object {
          "clearPdnCache": Object {
            "type": "boolean",
          },
          "ignore": Object {
            "items": Object {
              "type": "string",
            },
            "type": Array [
              "string",
              "array",
            ],
          },
          "ignoreModuleCache": Object {
            "type": "boolean",
          },
          "pdnHost": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "target": "dev.unbundle",
    },
  ],
]
`;
